src/python/pants/base/exceptions.py

```diff
@@ -5,6 +5,12 @@
 
 from typing import TYPE_CHECKING
 
+from pants.engine.internals.native_engine import EngineError as EngineError  # noqa: F401
+from pants.engine.internals.native_engine import (  # noqa: F401
+    IncorrectProductError as IncorrectProductError,
+)
+from pants.engine.internals.native_engine import IntrinsicError as IntrinsicError  # noqa: F401
```

Comment:
It's not clear to me why this uses new types, rather than using `NativeEngineFailure` (defined below). That doesn't necessarily mean that they should be merged, but it would be good to explain somewhere why they both exist.

`NativeEngineFailure` was introduced in #15979 to allow a `Failure` produced synchronously (i.e. raised) inside of a `@rule` to preserve its type information when it was caught at the `@rule` boundary. I _think_ that using the same type for `EngineError` would be fine, although I'm not certain.

Reply:
I'd like a clean class hierarchy here, but as there's a `failure` attr on the `NativeEngineFailure` I didn't feel like using it as a base for these other two, which are more like the "type information" it would hold, for passing back into `@rule`s.

Something like:

```mermaid
flowchart LR

subgraph rules
A[["@rule A"]]
B[["@rule B"]]
C[["@rule C"]]

A --> B --SomeException--> A
end

subgraph pantsd
engine --request--> A -- "NativeEngineFailure::Throw{SomeException}" --> engine
engine -- SomeException --> C --request--> engine
end


```

Reply:
Also want to follow up with changing all blanket `Exception` types thrown and instead use more dedicated error classes. Wasn't much of a point before, as all would just end up as fatal exits, but now that they'll be catchable it makes sense to classify them.

Reply:
> Wasn't much of a point before, as all would just end up as fatal exits, but now that they'll be catchable it makes sense to classify them.

I'd continue to recommend that we only use exceptions in particular circumstances, so would prefer to avoid introducing lots of new exception types at once: rather, for specific use cases.

Reply:
More like having the default be to throw `EngineError` rather than the base `Exception` by default. Then from there, if we need finer grained error types... sounds ok? (this in an attempt to not force the use of `catch Exception` unnecessarily)

Reply:
Yea, that's fine.

> this in an attempt to not force the use of `catch Exception` unnecessarily

Right. I'm just suggesting introducing the new exception type at the same time as the `catch`.